File: alggen_min_func.cpp
#include <iostream>
#include <cmath>
#include <bitset> //Sugestão para visualizar binário
#include <random>


using namespace std;

 //Setando constantes
const int NUM_PARES = 8; //Número de individuos. FIXME: Por ora funciona apenas com pares de pares
const int MAX_ITERACOES = 100;
const float TAXA_MUTACAO = 0.05f;


//Função usada: raizquadrada(x^3 + 2*(y^4))
inline double f(int x, int y){ return sqrt(pow(x, 3) + 2 * pow(y, 4)); }

int seletorPais(double *, random_device&);
void crossover(bitset<6>*, random_device&);
void mutacao(bitset<6>&, random_device&);
bitset<6> converteBitset(int, int);

//O algoritmo genético binário deve minimizar f(x) com x e y sendo valores no intervalo [0, 7]
//Serão 3 bits para cada entrada.
int main(){

    //Setando seed para criação de números aleatórios
    random_device rd;
    std::mt19937 mt(rd());
    std::uniform_int_distribution<int> dist(0, 7);
    std::uniform_int_distribution<int> roleta(0, 360);

    //Inicializando a população
    int pop[NUM_PARES][2]{};

    for (int i = 0; i < NUM_PARES; i++)
        for (int j = 0; j < 2; j++)
            pop[i][j] = dist(mt);
    
    //Avaliando os indivíduos
    double aval[NUM_PARES]{};
    double total_aval = 0.0;
    double melhor_aval = 100;   //melhor avaliação da iteração
    double melhor[2]{}; // melhor individuo

    for (int i = 0; i < NUM_PARES; i++){
        aval[i] = 1 + f(pop[i][0], pop[i][1]);  //Previnindo que a função de avaliação retorne valores negativos ou nulos, como no exemplo da sala
        if(aval[i] < melhor_aval){
            melhor_aval = aval[i];
            melhor[0] = pop[i][0];
            melhor[1] = pop[i][1];
        }
        total_aval += 1 / aval[i]; //Invertendo o valor já que o objetivo é minimizar a função
    }

    //Repartindo os pedaços da roleta
    double intervalos_roleta[NUM_PARES]{};
    for (int i = 0; i < NUM_PARES; i++){
        intervalos_roleta[i] = ((1.0 / aval[i]) / (total_aval)) *  360;  
        intervalos_roleta[i] = round(intervalos_roleta[i]);     //FIXME: Em alguns casos, os pedaços somam 359. Buscar uma forma de contornar o problema.
    }

    int i = 0;

    //Visualizando valores
    cout << "Populacao de individuos" << endl;
    for (int i = 0; i < NUM_PARES; i++)
        cout << "X: " << pop[i][0] << ", Y: " << pop[i][1] << endl;
    cout << endl;

    cout << "Avaliacao dos individuos sob f(x,y)" << endl;
    for (int i = 0; i < NUM_PARES; i++)
        cout << i + 1 << ":" << aval[i] << endl;
    cout << "Total: " << total_aval << endl << endl;

    cout << "Melhor individuo da geracao " << i << ": x = " << melhor[0] << ", y =" << melhor[1] << endl;
    cout << "------------------------------------------------------" << endl;
    do{
        //Selecionando pais
        int selecionados[NUM_PARES]{};
        for (int i = 0; i < NUM_PARES; i++){
            selecionados[i] = seletorPais(intervalos_roleta, rd);  
        }
        
        //Convertendo para bitset
        bitset<6> pais[NUM_PARES]{};

        for (int i = 0; i < NUM_PARES; i++){
            pais[i] = converteBitset(pop[selecionados[i]][0], pop[selecionados[i]][1]);
        }
        
        //Realizando o crossover
        crossover(pais, rd);

        //Reavaliando os indivíduos
        //TODO: Testar se isso aqui tá funcional
        total_aval = 0.0;

        for (int i = 0; i < NUM_PARES; i++){
            bitset<6> aux_x = (pais[i] >>= 3);
            bitset<6> aux_y = (pais[i] &= bitset<6>("000111"));
            
            pop[i][0] = aux_x.to_ulong();
            pop[i][1] = aux_y.to_ulong();
        }


        for (int i = 0; i < NUM_PARES; i++){
            aval[i] = 1 + f(pop[i][0], pop[i][1]);
            if(aval[i] < melhor_aval){
                melhor_aval = aval[i];
                melhor[0] = pop[i][0];
                melhor[1] = pop[i][1];
            }
            total_aval += 1 / aval[i]; 
        }

        //Repartindo os pedaços da roleta de novo
        for (int i = 0; i < NUM_PARES; i++){
            intervalos_roleta[i] = ((1.0 / aval[i]) / (total_aval)) *  360;  
            intervalos_roleta[i] = round(intervalos_roleta[i]);     //FIXME: Em alguns casos, os pedaços somam 359. Buscar uma forma de contornar o problema.
        }
        i++;

        //Visualizando valores
        cout << "Populacao de individuos" << endl;
        for (int i = 0; i < NUM_PARES; i++)
            cout << "X: " << pop[i][0] << ", Y: " << pop[i][1] << endl;
        cout << endl;

        cout << "Avaliacao dos individuos sob f(x,y)" << endl;
        for (int i = 0; i < NUM_PARES; i++)
            cout << i + 1 << ":" << aval[i] << endl;
        cout << "Total: " << total_aval << endl << endl;


        cout << "Melhor individuo da geracao " << i << ": x = " << melhor[0] << ", y =" << melhor[1] << endl;
        cout << "------------------------------------------------------" << endl;
    } while(melhor_aval > 1.0 && i < MAX_ITERACOES);
    
    //TODO: Procurar uma biblioteca para plotar o grafico em cpp. Sugerido o "matplotlib for cpp".  
    //TODO: Limpar e organizar o código.


 

    return 0;
}

int seletorPais(double * intervalos, random_device& rd){
    
    mt19937 mt(rd());
    uniform_int_distribution<int> roleta(0, 360);

    double limSup = intervalos[0];
    int num = roleta(mt);

    for (int i = 0; i < NUM_PARES; i++){
        if(num <= limSup)
            return i;
        else 
            limSup += intervalos[i + 1];
    }
    
    return -1;
}

bitset<6> converteBitset(int x, int y){
    bitset<6> binx(x), biny(y), ret(0);

    ret |= binx;
    ret <<= 3;
    ret |= biny;

    return ret;
}

void crossover(bitset<6> * pais, random_device& rd){
    
    mt19937 mt(rd());
    uniform_int_distribution<int> bitAlterado(0, 5);

    int num;

    for (int i = 0; i < NUM_PARES; i += 2){
        num = bitAlterado(mt);

        bitset<6> m1[2]{(0), (0)};   //Máscaras do primeiro pai
        bitset<6> m2[2]{(0), (0)};   //Máscaras do segundo pai

        switch (num){
        case 0:
                m1[0] |= 32, m1[1] |= 31;
                m2[0] |= 32, m2[1] |= 31;
                m1[0] &= pais[i];
                m1[1] &= pais[i];
                m2[0] &= pais[i + 1];
                m2[1] &= pais[i + 1];
                
                pais[i] = (m1[0] |= m2[1]);
                pais[i + 1] = (m2[0] |= m1[1]);
                mutacao(pais[i], rd);
                mutacao(pais[i + 1], rd);
            break;
        case 1:
                m1[0] |= 48, m1[1] |= 15;
                m2[0] |= 48, m2[1] |= 15;
                m1[0] &= pais[i];
                m1[1] &= pais[i];
                m2[0] &= pais[i + 1];
                m2[1] &= pais[i + 1];
                
                pais[i] = (m1[0] |= m2[1]);
                pais[i + 1] = (m2[0] |= m1[1]);
                mutacao(pais[i], rd);
                mutacao(pais[i + 1], rd);
            break;
        case 2:
                m1[0] |= 56, m1[1] |= 7;
                m2[0] |= 56, m2[1] |= 7;
                m1[0] &= pais[i];
                m1[1] &= pais[i];
                m2[0] &= pais[i + 1];
                m2[1] &= pais[i + 1];
                
                pais[i] = (m1[0] |= m2[1]);
                pais[i + 1] = (m2[0] |= m1[1]);
                mutacao(pais[i], rd);
                mutacao(pais[i + 1], rd);
            break;
        case 3:
                m1[0] |= 60, m1[1] |= 3;
                m2[0] |= 60, m2[1] |= 3;
                m1[0] &= pais[i];
                m1[1] &= pais[i];
                m2[0] &= pais[i + 1];
                m2[1] &= pais[i + 1];
                
                pais[i] = (m1[0] |= m2[1]);
                pais[i + 1] = (m2[0] |= m1[1]);
                mutacao(pais[i], rd);
                mutacao(pais[i + 1], rd);
            break;
        case 4:
                m1[0] |= 62, m1[1] |= 1;
                m2[0] |= 62, m2[1] |= 1;
                m1[0] &= pais[i];
                m1[1] &= pais[i];
                m2[0] &= pais[i + 1];
                m2[1] &= pais[i + 1];
                
                pais[i] = (m1[0] |= m2[1]);
                pais[i + 1] = (m2[0] |= m1[1]);
                mutacao(pais[i], rd);
                mutacao(pais[i + 1], rd);
            break;
        case 5:
                m1[0] |= 63, m1[1] |= 0;
                m2[0] |= 63, m2[1] |= 0;
                m1[0] &= pais[i];
                m1[1] &= pais[i];
                m2[0] &= pais[i + 1];
                m2[1] &= pais[i + 1];
                
                pais[i] = (m1[0] |= m2[1]);
                pais[i + 1] = (m2[0] |= m1[1]);
                mutacao(pais[i], rd);
                mutacao(pais[i + 1], rd);
            break;
        default:
            cout << "Algo deu errado no crossover..." << endl;
            break;
        }
        
    }
}

void mutacao(bitset<6> & individuo, random_device& rd){
    
    mt19937 mt(rd());
    uniform_int_distribution<int> mutar(0, 99);

    if(mutar(mt) < (TAXA_MUTACAO * 100)){
        uniform_int_distribution<int> bitAlterado(0, 5);
        int num = bitAlterado(mt);
        individuo.flip(num);    //Inverte o bit na posição escolhida, da direita pra esquerda
    }
}

File: alggen_min_func.py
import numpy as np
import matplotlib.pyplot as plt
import random

# Setando constantes
NUM_PARES = 8
MAX_ITERACOES = 50
TAXA_MUTACAO = 0.05

# Função usada: raizquadrada(x^3 + 2*(y^4))
def f(x, y):
    return np.sqrt(np.power(x, 3) + 2 * np.power(y, 4))

def seletor_pais(intervalos):
    num = random.randint(0, 360)
    lim_sup = intervalos[0]

    for i in range(NUM_PARES):
        if num <= lim_sup:
            return i
        else:
            lim_sup += intervalos[i + 1]

    return -1

def converte_bitset(x, y):
    bin_x = np.binary_repr(x, width=3)
    bin_y = np.binary_repr(y, width=3)
    bin_str = bin_x + bin_y
    return int(bin_str, 2)

def crossover(pais):
    num = random.randint(0, 5)

    for i in range(0, NUM_PARES, 2):
        m1 = [0, 0]  # Máscaras do primeiro pai
        m2 = [0, 0]  # Máscaras do segundo pai

        if num == 0:
            m1[0] = 32
            m1[1] = 31
            m2[0] = 32
            m2[1] = 31
        elif num == 1:
            m1[0] = 48
            m1[1] = 15
            m2[0] = 48
            m2[1] = 15
        elif num == 2:
            m1[0] = 56
            m1[1] = 7
            m2[0] = 56
            m2[1] = 7
        elif num == 3:
            m1[0] = 60
            m1[1] = 3
            m2[0] = 60
            m2[1] = 3
        elif num == 4:
            m1[0] = 62
            m1[1] = 1
            m2[0] = 62
            m2[1] = 1
        elif num == 5:
            m1[0] = 63
            m1[1] = 0
            m2[0] = 63
            m2[1] = 0

        m1[0] &= pais[i]
        m1[1] &= pais[i]
        m2[0] &= pais[i + 1]
        m2[1] &= pais[i + 1]

        pais[i] = m1[0] | m2[1]
        pais[i + 1] = m2[0] | m1[1]
        mutacao(pais[i])
        mutacao(pais[i + 1])

def mutacao(individuo):
    x = random.randint(0, 4)
    if x < TAXA_MUTACAO:
        bit_alterado = random.randint(0, 5)
        individuo ^= (1 << bit_alterado)

# Setando seed para criação de números aleatórios
pop = np.array([random.randint(0, 7) for _ in range(2 * NUM_PARES)]).reshape(NUM_PARES, 2)

# Avaliando os indivíduos
aval = np.zeros(NUM_PARES)
total_aval = 0.0
melhor_aval = 100
melhor = np.zeros(2)

melhores_x = [melhor[0]]
melhores_y = [melhor[1]]

for i in range(NUM_PARES):
    aval[i] = 1 + f(pop[i][0], pop[i][1])
    if aval[i] < melhor_aval:
        melhor_aval = aval[i]
        melhor[0] = pop[i][0]
        melhor[1] = pop[i][1]
        melhores_x.append(melhor[0])
        melhores_y.append(melhor[1])
    total_aval += 1 / aval[i]

# Repartindo os pedaços da roleta
intervalos_roleta = np.zeros(NUM_PARES)
for i in range(NUM_PARES):
    intervalos_roleta[i] = ((1.0 / aval[i]) / total_aval) * 360
    intervalos_roleta[i] = round(intervalos_roleta[i])

i = 0
print(f"Melhor individuo da geracao {i}: x = {melhor[0]}, y = {melhor[1]}")
while melhor_aval > 1.0 and i < MAX_ITERACOES:
    # Selecionando pais
    selecionados = [seletor_pais(intervalos_roleta) for _ in range(NUM_PARES)]

    # Convertendo para bitset
    pais = [converte_bitset(pop[selecionado][0], pop[selecionado][1]) for selecionado in selecionados]

    # Realizando o crossover
    crossover(pais)

    # Reavaliando os indivíduos
    total_aval = 0.0

    for j in range(NUM_PARES):
        aux_x = pais[j] >> 3
        aux_y = pais[j] & 7

        pop[j][0] = aux_x
        pop[j][1] = aux_y

    for j in range(NUM_PARES):
        aval[j] = 1 + f(pop[j][0], pop[j][1])
        if aval[j] < melhor_aval:
            melhor_aval = aval[j]
            melhor[0] = pop[j][0]
            melhor[1] = pop[j][1]
            melhores_x.append(melhor[0])
            melhores_y.append(melhor[1])
        total_aval += 1 / aval[j]

    # Repartindo os pedaços da roleta de novo
    for j in range(NUM_PARES):
        intervalos_roleta[j] = ((1.0 / aval[j]) / total_aval) * 360
        intervalos_roleta[j] = round(intervalos_roleta[j])

    i += 1
    print(f"Melhor individuo da geracao {i}: x = {melhor[0]}, y = {melhor[1]}")

# Visualizando valores
print("Populacao de individuos")
for i in range(NUM_PARES):
    print(f"X: {pop[i][0]}, Y: {pop[i][1]}")

print("\nAvaliacao dos individuos sob f(x,y)")
for i in range(NUM_PARES):
    print(f"{i + 1}: {aval[i]}")

print(f"Total: {total_aval}")

melhores_x = [melhor[0]]
melhores_y = [melhor[1]]

plt.plot(i, melhores_x, label='Melhor X', marker='o', linestyle='-')
plt.plot(i, melhores_y, label='Melhor Y', marker='o', linestyle='-')
plt.title('Evolução dos Melhores Indivíduos')
plt.xlabel('Geração')
plt.ylabel('Melhor Indivíduo')
plt.legend()
plt.grid(True)
plt.show()